File: banking_system.c
/*
1. Create an online banking system with the following features:
* Users must be able to log in with a username and password.
* If the user enters the wrong credentials three times, the system must lock them out.
* The initial balance in the bank account is $2000.
* The system must allow users to deposit, withdraw, view, and transfer money.
* The system must display a menu for users to perform transactions.
*/
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct user{
    char username[256];
    char password[256];
    float balance;
}user;

int main(){
    //initializing variables
    int log_in_counter = 0;
    int counter;
    int user_counter=0;
    int option;
    int second_option;
    float amount;
    int menu1 = 1;
    int menu2;
    int aux;
    int flag_log_in;
    int flag_transaction;
    user users[256];
    char *ptruser, *ptrpassw;

    while(menu1 == 1){
        printf("Banking System\n1- Create account\n2- Log in\n3- Exit\n");
        scanf("%i",&option);
        switch (option)
        {
        //Creating an account
        case 1:
        //Defining an address in memory for each pointer 
            ptruser = malloc(sizeof(char));
            ptrpassw = malloc(sizeof(char));
            printf("\nEnter a  new username, no blank spaces:");
            scanf("%s",ptruser);
            printf("\nEnter a new password, no blank spaces:");
            scanf("%s",ptrpassw);
            strcpy(users[user_counter].username,ptruser);
            strcpy(users[user_counter].password,ptrpassw);
            users[user_counter].balance = 2000;
            user_counter ++;
            printf("\nNew user %s , new password %s, balance:%f\nAccount created successfully\n\n",users[user_counter-1].username,users[user_counter-1].password,users[user_counter-1].balance);
            free(ptruser);
            free(ptrpassw);
            ptruser = NULL;
            ptrpassw = NULL;
            break;
        //Log in
        case 2:
            if(user_counter==0){
                printf("there no users in the system, please create an user\n");
                break;
            }else{
                ptruser = malloc(sizeof(char));
                ptrpassw = malloc(sizeof(char));
                printf("\nEnter your credentials\nUsername:");
                scanf("%s",ptruser);
                printf("\nEnter yous password\n");
                scanf("%s",ptrpassw);
                counter = 0;
                flag_log_in=0;
                while (counter < user_counter) {
                    if(log_in_counter < 3){
                        if(0 == strcmp(ptruser,users[counter].username) && 0 == strcmp(ptrpassw,users[counter].password)){
                            free(ptruser);
                            ptruser = NULL;
                            flag_log_in=1;
                            menu2=1;
                            printf("\nValid Credentials\n\n");
                            while(menu2 == 1){
                                printf("Select an operation:\n1-Withdraw\n2-Deposit\n3-View balance\n4-Transfer\n5-Exit\n");
                                scanf("%i",&second_option);
                                switch (second_option){
                                    //Withdraw
                                    case 1:
                                        printf("\nHow much do you want to withdraw?:");
                                        scanf("%f",&amount);
                                        if(amount > users[counter].balance || amount <= 0){
                                        printf("Invalid amount, try again\n\n");
                                        }else{
                                            users[counter].balance = users[counter].balance - amount;
                                            printf("Withdraw Successful\n\n");
                                        }
                                        break;
                                    //Desposit
                                    case 2:
                                        printf("\nHow much do you want to deposit?");
                                        scanf("%f",&amount);
                                        if(amount<=0){
                                            printf("Invalid amount, try again\n");
                                        }else{
                                            users[counter].balance = users[counter].balance + amount;
                                            printf("Deposit successful\n\n");
                                        }
                                        break;
                                    //View Balance
                                    case 3:
                                        printf("Your Balance is %f\n\n",users[counter].balance);
                                        break;
                                    //Transfer
                                    case 4:
                                        if(user_counter<2){
                                            printf("Too few users, create more accounts\n\n");
                                            break;
                                        }else{
                                            ptruser = malloc(sizeof(char));
                                            aux = 0;
                                            printf("Transaction Menu\nWrite the user ID to which transfer the money(it's the same as their username)\nUser ID:");
                                            scanf("%s",ptruser);
                                            flag_transaction=0;
                                            while(aux < user_counter){
                                                if(0 == strcmp(ptruser,users[aux].username)){
                                                    flag_transaction=1;
                                                    printf("How much money do you want to transfer?:");
                                                    scanf("%f",&amount);
                                                    if(amount > users[counter].balance || amount <=0){
                                                        printf("Invalid amount\n\n");
                                                        break;
                                                    }else{
                                                        printf("\nSuccessfull transtion\n\n");
                                                        users[aux].balance = users[aux].balance + amount;
                                                        users[counter].balance = users[counter].balance - amount;
                                                    }
                                                }
                                                aux++;                                       
                                            }
                                            if(flag_transaction==0){
                                                printf("\nWrong ID, try again\n\n");
                                            }
                                            free(ptruser);
                                            ptruser = NULL;
                                        }
                                        break;
                                    //Exit
                                    case 5:
                                        menu2=0;
                                        break;
                                    default:
                                        printf("Invalid option, try again\n\n");
                                        break;
                                }
                            }
                            break;
                        }
                    }
                    //Locking out the user 
                    counter++;
                }
                free(ptruser);
                free(ptrpassw);
                ptrpassw= NULL;
                ptruser = NULL;
                if(flag_log_in==0){
                    printf("Wrong Credentials\n\n");
                    log_in_counter++;
                }
                if(log_in_counter==3){
                    printf("Too many attempts, Locking the System\n\n");
                    menu1=0;
                }
                counter = 0;
                break;
            }
        case 3:
            menu1 = 0;
            break;
        default:
            printf("\nInvalid option, try again\n\n");
            break;
        }
    }
    return 0;
}

File: banking_system.py
'''
Create an online banking system with the following features:

* Users must be able to log in with a username and password.
* If the user enters the wrong credentials three times, the system must lock them out.
* The initial balance in the bank account is $2000.
* The system must allow users to deposit, withdraw, view, and transfer money.
* The system must display a menu for users to perform transactions. 
'''
class User:
    userid:str
    password:str
    balance:int

    def __init__(self, userid, password, balance=2000):
        self.userid = userid
        self.password = password
        self.balance = balance       

user_credentials = [User(userid="user", password="password"), User(userid="Mayber", password="codingchallenge")]

    
def create_account(userid,password):
    return {"Error": "El usuario ya existe"} if search_user(userid) else user_credentials.update({userid:password})
def menu(userid):
    control = "n"
    
    while control == "n":
        print("Welcome.\n")
        print("1. ver saldo.")
        print("2. depositar dinero.")
        print("3. retirar dinero.")
        print("4. transferir dinero.")
        opcion= int(input())

        match(opcion):
            case 1:
                print(view(userid))
            case 2:
                sum= int(input("ingrese monto a depositar.\n"))
                deposit(userid, sum)
            case 3:
                sum= int(input("ingrese monto a retirar.\n"))
                withdraw(userid,sum)
            case 4:
                dest_userid=input("ingrese el ID del usuario destino.\n")
                sum = int(input("ingrese monto a transferir.\n"))
                transfer(userid, sum, dest_userid)
            case _:
                print("Please insert a Valid option.")

        control=input("Desea salir? y/n\n")
        
        if control == 'y':
            print("Hasta luego. Vuelva pronto!")
    
def search_user(userid:str)->User | bool:
    '''
    hace una búsqueda del usuario en las credenciales almacenadas si el usuario existe.
    
    :param: userid:str - ID proporcionado por el usuario.
    :return: object or error - objeto resultante contiene datos del usuario.

    '''
    user= filter(lambda user: user.userid == userid, user_credentials)
    try:
        return list(user)[0]
    except:
        return False
    
def login(userid:str, password:str)->bool:
    '''
    valida que las credenciales pertenezcan al usuario.

    :param: userid, password: str - ID y Password proporcionados por el usuario.
    :return: bool - son correctas ambas credenciales?
    '''
    user= search_user(userid)
    if not user: return False
    if userid == user.userid and password == user.password: return True 
    

def deposit(userid:str,sum):
    '''sumar cantidad actual mas deposito'''
    user=search_user(userid)
    if type(user) == User:
        user.balance += sum
        print("Success.\n")
        return 

def withdraw(userid,sum):
    '''restar cantidad actual menos deposito'''
    user=search_user(userid)
    if type(user) == User:
        if sum > user.balance:
            print("Not enough funds.")
            return
        user.balance -= sum
        print (f"success.\n {view(userid)}")
    return

def view(userid):
    '''mostrar balance actual.
    :param: userid: str - usuario actual.
    :return: str - balance de cuenta.
    '''
    user=search_user(userid)
    if type(user) == User:
        return f"Balance actual: {user.balance}"
    
def transfer(userid:str,sum:int, dest_userid:str):
    '''
    validar userid del cliente destino, 
    validar que la cantidad a transferir sea menor o igual a la disponible,
    restar del balance del usuario actual la cantidad a transferir.

    :param: userid, dest_userid: str, sum: int - id del usuario que envia, 
    usuario que recibe y suma a enviar.
    :return: str - success or failure.
    '''
    
    if type(search_user(userid)) == User and type(search_user(dest_userid)) == User:
        withdraw(userid,sum)
        deposit(dest_userid,sum)
        return
    else: 
        print("error!")
        return
    
        
    

def main():
    print("bienvenido! Por favor ingrese sus datos.")
    
    userid = input("ingrese su ID: ")
    password = input("ingrese su contraseña: ")
    limit = 2

    while not login(userid, password)and limit >0:
        print("wrong credentials!")
        limit -=1
        userid = input("ingrese su ID: ")
        password = input("ingrese su contraseña: ")
    if limit == 0:
        print("sorry, you're suspended")
    else:  
        menu(userid)


if __name__ == "__main__":
    main()
        